classes/order/OrderDetail.php

```diff
@@ -533,28 +533,45 @@ protected function setVirtualProductInformation($product)
     }
 
     /**
-     * Check the order status.
+     * Updates product quantity in stock, according to order status.
      *
      * @param array $product
-     * @param int $id_order_state
+     * @param int $orderStateId
      */
-    protected function checkProductStock($product, $id_order_state)
+    protected function updateProductQuantityInStock($product, $orderStateId): void
     {
-        if ($id_order_state != Configuration::get('PS_OS_CANCELED') && $id_order_state != Configuration::get('PS_OS_ERROR')) {
-            $update_quantity = true;
-            if (!StockAvailable::dependsOnStock($product['id_product'])) {
-                $update_quantity = StockAvailable::updateQuantity($product['id_product'], $product['id_product_attribute'], -(int) $product['cart_quantity'], $product['id_shop'], true);
-            }
-
-            if ($update_quantity) {
-                $product['stock_quantity'] -= $product['cart_quantity'];
-            }
-
-            if ($product['stock_quantity'] < 0 && Configuration::get('PS_STOCK_MANAGEMENT')) {
-                $this->outOfStock = true;
-            }
-            Product::updateDefaultAttribute($product['id_product']);
+        $dismissOrderStateIds = Configuration::getMultiple([
+            'PS_OS_CANCELED',
+            'PS_OS_ERROR',
+        ]);
+        if (in_array($orderStateId, $dismissOrderStateIds)) {
+            return;
+        }
+        if (!StockAvailable::dependsOnStock($product['id_product'])) {
+            $orderState = new OrderState($orderStateId, $this->id_lang);
+            $isQuantityUpdated = StockAvailable::updateQuantity(
+                $product['id_product'],
+                $product['id_product_attribute'],
+                -(int) $product['cart_quantity'],
+                $product['id_shop'],
+                // Add stock movement only if product is not virtual and order state is flagged as shipped
+                empty($product['is_virtual']) && true === (bool) $orderState->shipped,
+                [
+                    'id_order' => $this->id_order,
+                    // Only one stock movement reason fits a new order creation
+                    'id_stock_mvt_reason' => Configuration::get('PS_STOCK_CUSTOMER_ORDER_REASON'),
+                ]
+            );
+        } else {
+            $isQuantityUpdated = true;
+        }
+        if ($isQuantityUpdated === true) {
+            $product['stock_quantity'] -= $product['cart_quantity'];
+        }
+        if ($product['stock_quantity'] < 0 && Configuration::get('PS_STOCK_MANAGEMENT')) {
```

Comment:
I'm maybe missing something (and out of scope) but shouldn't the product be out of stock when `stock_quantity <= 0` instead of `< 0`?

Reply:
Good call, but I think it's ok this way If the remaining stock is strictly 0 it means the order just got the last left items 

I believe the `$this->outOfStock` parameter is here to indicate that this order is explicitly using some nonexisting stock (probably linked the the availability to order Out of stock products)

Reply:
`outOfStock` property seems only read by payment module from `OrderDetail::getStockState()`, in order to persist order history and to send email to customer.
https://github.com/PrestaShop/PrestaShop/blob/d1d4982ee74c70ac331de2b76cb00c84912d820c/classes/PaymentModule.php#L585